File: Kotlin/sort/RadixSort/src/RadixSort.kt
// Function to Implement Radix Sort
// The radix sorting algorithm is an integer sorting algorithm, 
// that sorts by grouping numbers by their individual digits (or by their radix).
// It uses each radix/digit as a key, and implements counting sort or 
// bucket sort under the hood in order to do the work of sorting.

// Takes an IntArray as arguments and returns a sorted IntArray to the caller function
fun radixSort(original: IntArray): IntArray {
    // Array needs to be mutable
    var old = original
    for (shift in 31 downTo 0) {
        val tmp = IntArray(old.size)
        // The number of 0s
        var j = 0
        // Move the 0s to the new array, and the 1s to the old one
        for (i in 0 until old.size) {
            // If there is a 1 in the bit we are testing, the number will be negative
            val move = (old[i] shl shift) >= 0
            // If this is the last bit, negative numbers are actually lower
            val toBeMoved = if (shift != 0) move else !move
            if (toBeMoved) {
                tmp[j++] = old[i]
            }
            else {
                // It's a 1, so stick it in the old array for now
                old[i - j] = old[i]
            }
        }
        // Copy over the 1s from the old array
        for (i in j until tmp.size) {
            tmp[i] = old[i - j]
        }
        old = tmp
    }
    return old
}

fun main(args: Array<String>) {
    print("Enter N: ");
    val n= readLine()!!.toInt();
    println("Enter array of N integers: ");
    val arrays = IntArray(n) { readLine()!!.toInt() }
    val array= radixSort(arrays);
    print("Array after Radix Sort is: ")
    for(i in array){
        print("$i ")
    }
}

/*
Time Complexity: O(log n)
Sample Input:
Enter N: 5
Enter array of N integers:
3
-1
2
3
4

Sample Output:
Array after Radix Sort is: -1 2 3 3 4
 */

File: C-Sharp/sort/RadixSort.cs
using System;

/*  Radix sort is a sorting technique that sorts the elements by 
    first grouping the individual digits of the same place value. 
    Then, sort the elements according to their increasing/decreasing order.
*/

 class RadixSort{
  static void radixSort(int[] arr){
    int size = arr.Length;
    int max_element = arr[0];
    
    for (int i=1; i<size; i++){  
      if(max_element < arr[i])
        max_element = arr[i];
    }

    for (int x = 1; max_element/x > 0; x *= 10){ 
      countingSort(arr, x); 
      }
  }

 
  static void print_arr(int[] arr){ 
    int size = arr.Length; 
   
    Console.Write("\n"); 
  } 
  
    static void countingSort(int[] arr, int x){   
    int size = arr.Length;    
    int[] new_arr = new int[size];
    
    int[] fr = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

    for(int i = 0; i < size; i++){
      fr[(arr[i]/x)%10]++;
      }

    for (int i = 1; i < 10; i++){ 
      fr[i] += fr[i - 1];     
      }

    for (int i = size - 1; i >= 0; i--){ 
      new_arr[fr[(arr[i]/x)%10] - 1] = arr[i]; 
      fr[(arr[i]/x)%10]--; 
    } 
  
    for (int i = 0; i < size; i++){ 
      arr[i] = new_arr[i]; 
     } 
  }


  public static void Main(string[] args){
    int n = Convert.ToInt32(Console.ReadLine());
    int[] a = new int[n];
   
   int i,j;
    for(i=0; i<n;i++){
        Console.Write("Element {0} : ", i);
        a[i] = Convert.ToInt32(Console.ReadLine());
    }
    Console.Write("Unsorted array: ");
    for (j=0; j<n; j++){
      Console.Write(a[j] + " "); 
    } 
      
    Console.Write("\n");
    
    radixSort(a);
    
    Console.Write("Sorted array: ");
     for (j=0; j<n; j++){
      Console.Write(a[j] + " "); 
      } 
  }
}

/* Input
    6                                                                                         
    Element 0 : 87                                                                            
    Element 1 : 12                                                                            
    Element 2 : 4                                                                             
    Element 3 : 76                                                                            
    Element 4 : 1                                                                             
    Element 5 : 23 
*/

/* Output
    Unsorted array: 87 12 4 76 1 23 
    Sorted array: 1 4 12 23 76 87 
*/

/* Time Complexity = O(d(n+k)) 
        where k= largest element in the array
              n= number of elements in the array
              d = base of numbers(decimal base system =10)
*/


